# NGS Summer 2016 - Analyzing Next-Generation Sequencing Data

August 8 - August 19th, 2016
Kellogg Biological Station, Michigan State University

Please sign up for [the announcements mailing](http://lists.idyll.org/listinfo/ngs-course-announce) list if you are interested in taking this class now, or in the future.

### Instructors: 
Prof. Meg Staton and Prof. Matt MacManes along with Dr Torsten Seemann, Shaun Jackman and others.  


###  News
Feb X, 2016 - application form is now open! We will close the application period on April 07th (midnight EST) and notify applicants of decisions on April 17th.


### Course Description
This intensive two week summer course will introduce attendees with a strong biology background to the practice of analyzing short-read sequencing data from Illumina and other next-gen platforms. The first week will introduce students to computational thinking and large-scale data analysis on UNIX platforms. The second week will focus on mapping, assembly, and analysis of short-read data for resequencing, ChIP-seq, and RNAseq.

No prior programming experience is required, although familiarity with some programming concepts is helpful, and bravery in the face of the unknown is necessary. 2 years or more of graduate school in a biological science is strongly suggested. Faculty, postdocs, and research staff are more than welcome!

Students will gain practical experience in:

- Python and bash shell scripting
- cloud computing/Amazon EC2
- basic software installation on UNIX
- installing and running bowtie, Trinity, Velvet, SPAdes and other bioinformatics tools.
- querying mappings and evaluating assemblies
- Materials from previous courses are available at http://angus.readthedocs.org/ under a Creative Commons/full use+reuse license.

You can read a blog post about the 2015 course [here](http://ivory.idyll.org/blog/tag/ngs-course.html).

### What will students learn?
By the end of the course, students will be able to assemble transcriptomes, identify variants, and estimate transcript abundance (from mRNAseq data). They will also know how to transfer large data sets between computers, run extended analyses while asleep, execute and modify existing Python scripts, and otherwise effectively make use of basic computational resources.

### Location, dates, and course structure.
The course will be held at the W.K. Kellogg Biological Station on Gull Lake (http://www.kbs.msu.edu/) in western Michigan from noon on Monday, August 8th through 2pm on Friday, August 19th. Morning and afternoon lectures will be interspersed with practical hands-on labs. Room and board will be provided on-site (see enrolling, below). Sunday, August 16th, will be a day of rest & relaxation. Students should expect to arrive at KBS by Monday, August 8th at noon and stay at KBS through Friday, August 19th at 2pm. KBS is several hours away from the nearest large airports (DTW, GRR, LAN) and students coming from out of state should plan on arriving on Sunday.

### Applying for the course
An application is required, and we can accommodate only 24 attendees. We absolutely welcome "students" of any career stage, including postdocs and faculty. Applicants are not selected on a first-come-first-serve basis, and we try to create a balanced class of participants.

*Note that some family housing is available*, and there may be child care opportunities as well by special arrangement. Please contact Matt or Meg _after_ you have been admitted to discuss details!

### Tuition, course cost, and enrolling
All students must pay for on-site room and board (see http://conference.kbs.msu.edu/conference-center/plan-conference/) prior to arrival. It should be about $650. We will make arrangements for housing *after* choosing students; please do not contact KBS directly. We will also charge a course fee of 250/person.

Applicants must commit to spending both weeks on-site! The middle Sunday will be entirely free.

### Taking the course for-credit
There is no formal option for taking the course for credit at MSU or elsewhere.

### Supplies
Please plan to bring a laptop (Mac OS X, Windows, or Linux are all OK). We will not have any additional laptops available this year.

### Lodging and family
In 2016, we plan to have housing available for family groups. Child care may be available by arrangement.

### Instructors
Matt MacManes (http://genomebio.org/) holds a Ph.D. in Integrative Biology from UC Berkeley, and has worked in behavioral ecology, population genetics, and more recently transcriptomics and genomics. He is currently an Assistant Professor in Molecular Cellular and Biomedical Sciences at The University of New Hampshire, where his lab works on functional and population genomics in non-model organisms including desert mice, ladybugs, clams, corals, and doves.

Meg Staton (http://agbioinfo.utk.edu/) is an Assistant Professor in the Department of Entomology and Plant Pathology at the University of Tennessee, Knoxville. She specializes in bioinformatics and genomics of plants and her lab helps build the software Tripal, runs genomic data web portals, and analyzes a wide variety of genomics data.

### Board of Advisors
Human genetics and genomics: Kevin White, James and Karen Frank Family Professor at University of Chicago.

Animal genetics and genomics: Paul W. Sternberg, Thomas Hunt Morgan Professor of Biology at the California Institute of Technology; member of the National Academy of Sciences.

Microbial evolution and resequencing: Richard E. Lenski, Hannah Distinguished Professor at Michigan State University; member of the National Academy of Sciences.

Plant genetics and genomics: Robin Buell, Associate Professor at Michigan State University.

Microbial population sequencing and metagenomics: James M. Tiedje,
Professor, Michigan State University; member of the National Academy of Sciences.

Bioinformatics and genomics: Lincoln Stein, Platform Leader, Informatics and Bio-Computing, Ontario Institute for Cancer Research; Professor, Cold Spring Harbor Laboratory.

### Funding
This course uses materials developed under AFRI Competitive Grant no. 2010-65205-20361 from the USDA NIFA, and is supported by Grant Number R25HG006243 from the National Human Genome Research Institute, both to C. Titus Brown.
